h/static/scripts/group-forms/utils/unsaved-changes.ts

```diff
@@ -0,0 +1,43 @@
+import { useEffect } from 'preact/hooks';
+
+/** Count of components with unsaved changes. */
+let unsavedCount = 0;
+
+function preventUnload(e: BeforeUnloadEvent) {
+  e.preventDefault();
+  e.returnValue = '';
+}
+
+/**
+ * Return true if any active components have indicated they have unsaved changes
+ * using {@link useUnsavedChanges}.
+ */
+export function hasUnsavedChanges() {
+  return unsavedCount > 0;
+}
+
+/**
+ * Hook that registers the current component as having unsaved changes that
+ * would be lost in the event of a navigation.
+ *
+ * @param hasUnsavedChanges - True if current component has unsaved changes
+ * @param window_ - Test seam
+ */
+export function useUnsavedChanges(hasUnsavedData: boolean, window_ = window) {
+  useEffect(() => {
+    if (hasUnsavedData) {
+      unsavedCount += 1;
+      if (unsavedCount === 1) {
+        window_.addEventListener('beforeunload', preventUnload);
+      }
+    }
+    return () => {
+      if (hasUnsavedData) {
+        unsavedCount -= 1;
+        if (unsavedCount === 0) {
+          window_.removeEventListener('beforeunload', preventUnload);
+        }
+      }
+    };
```

Comment:
If I'm getting this right, all the components that call this hook passing `hasUnsavedData = true` will increment the count in 1, but only the first time the event handler will be registered (which makes sense, as it's a global page-level event).

And at the same time, the listener will not be removed until all the components that called the hook have been unmounted, which again makes sense.

Perhaps we could simplify the conditions though, as it took me a moment to wrap my head around the nested conditions:

```suggestion
    if (!hasUnsavedData) {
      return () => {};
    }
    
    unsavedCount += 1;
    // Register even listener only for the first component indicating there are unsaved changes
    if (unsavedCount === 1) {
      window_.addEventListener('beforeunload', preventUnload);
    }
    
    return () => {
      unsavedCount -= 1;
      // Remove the listener only when there are no components left with unsaved changes
      if (unsavedCount === 0) {
          window_.removeEventListener('beforeunload', preventUnload);
      }
    };
```